### PR TITLE
Retain name in PeriodIndex resample

### LINKED ISSUE #12769: BUG: Resample loses PeriodIndex name
#### Code Sample, a copy-pastable example if possible
```python

In [17]: series=pd.Series(range(10), pd.period_range(start='2000', periods=10, name='date', freq='M'))

In [18]: series
Out[18]: 
date
2000-01    0
2000-02    1
2000-03    2
2000-04    3
2000-05    4
2000-06    5
2000-07    6
2000-08    7
2000-09    8
2000-10    9
Freq: M, dtype: int64

In [19]: series.index
Out[19]: 
PeriodIndex(['2000-01', '2000-02', '2000-03', '2000-04', '2000-05', '2000-06',
             '2000-07', '2000-08', '2000-09', '2000-10'],
            dtype='int64', name='date', freq='M')
# Note name='date'

In [20]: series.resample('D').pad().index
Out[20]: 
PeriodIndex(['2000-01-01', '2000-01-02', '2000-01-03', '2000-01-04',
             '2000-01-05', '2000-01-06', '2000-01-07', '2000-01-08',
             '2000-01-09', '2000-01-10',
             ...
             '2000-10-22', '2000-10-23', '2000-10-24', '2000-10-25',
             '2000-10-26', '2000-10-27', '2000-10-28', '2000-10-29',
             '2000-10-30', '2000-10-31'],
            dtype='int64', length=305, freq='D')

# now no name
```
#### Expected Output

```python
Out[20]: 
PeriodIndex(['2000-01-01', '2000-01-02', '2000-01-03', '2000-01-04',
             '2000-01-05', '2000-01-06', '2000-01-07', '2000-01-08',
             '2000-01-09', '2000-01-10',
             ...
             '2000-10-22', '2000-10-23', '2000-10-24', '2000-10-25',
             '2000-10-26', '2000-10-27', '2000-10-28', '2000-10-29',
             '2000-10-30', '2000-10-31'],
            dtype='int64', length=305, freq='D', name='date')
```

#### output of `pd.show_versions()`

`Out[21]: '0.18.0'`

### PR DESCRIPTION
**Pull request by max-sixty**
- [x] closes #12769
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tseries/resample.py [hunk 1]
```diff
@@ -652,9 +652,6 @@ def _convert_obj(self, obj):
             # Cannot have multiple of periods, convert to timestamp
             self.kind = 'timestamp'
 
-        if not len(obj):
-            self.kind = 'timestamp'
-
```
Comment (MaximilianR) [inline pandas/tseries/resample.py]: I'm not sure what the goal of this was. It made [this test](https://github.com/pydata/pandas/pull/12771/files#diff-f89a17cef9382e5fafdc58604f1744c0R2258) I added fail

File: pandas/tseries/tests/test_timeseries.py [hunk 2]
```diff
@@ -2249,15 +2245,26 @@ def test_concat_datetime_datetime64_frame(self):
...
         s[10:30] = np.nan
         expected = Series([34.5, 79.5], index=[Period(
             '2013-01-01 00:00', 'T'), Period('2013-01-01 00:01', 'T')])
+        expected.index.name = 'idx'
         result = s.to_period().resample('T', kind='period').mean()
         assert_series_equal(result, expected)
         result2 = s.resample('T', kind='period').mean()
         assert_series_equal(result2, expected)
 
+    def test_empty_period_index_resample(self):
```
Comment (MaximilianR) [inline pandas/tseries/tests/test_timeseries.py]: This now tests [this](https://github.com/pydata/pandas/pull/12771/files#diff-e21e44fc4ab14efc2f09e0dacdd79fc6R676), which was untested

File: pandas/tseries/tests/test_timeseries.py [hunk 3]
```diff
@@ -1,42 +1,38 @@
...
-import pandas.compat as compat
-import pandas.core.common as com
-from pandas import concat
-from pandas import _np_version_under1p8
-
-from numpy.testing.decorators import slow
+from pandas.tslib import iNaT
+from pandas.util.testing import (
+    assert_frame_equal, assert_series_equal, assert_almost_equal,
+    _skip_if_has_locale, assert_index_equal)
```
Comment (MaximilianR) [inline pandas/tseries/tests/test_timeseries.py]: Cleaned this section up a bit, partly because PyCharm started on my behalf

**Comment by jreback** [comment]
pls don't change stuff just because py charm does
use targeted changes

**Comment by MaximilianR** [comment]
I would remove it if I didn't think it made it better / cleaner. Happy to if you don't think it does...

File: doc/source/whatsnew/v0.18.1.txt [hunk 4]
```diff
@@ -192,7 +192,7 @@ Bug Fixes
 
 
 - Bug in ``CategoricalIndex.get_loc`` returns different result from regular ``Index`` (:issue:`12531`)
-
+- Bug in ``PeriodIndex.resample`` where name not propagated
```
Comment (jreback) [inline doc/source/whatsnew/v0.18.1.txt]: add the issue number

File: pandas/tseries/resample.py (hunk 1 above)
Comment (jreback) [inline pandas/tseries/resample.py]: the ``kind`` argument is quite tricky. We really should remove it, but that's a whole other project. The issue is that some of the routines are expecting ``DatetimeIndex`` so we give it to them even though we are a ``PeriodIndexResampelr``. This all existed before the changed in 0.18.0, but was much trickier to deal with as it was function based. So should be possible to fix now.

File: pandas/tseries/tests/test_timeseries.py (hunk 3 above)
Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: its fine. problem is these things are cleaned differntly based on settings. 

I guess it reordered it a bit more logically, ok then.

File: pandas/tseries/tests/test_timeseries.py [hunk 5]
```diff
@@ -2249,15 +2245,26 @@ def test_concat_datetime_datetime64_frame(self):
     def test_period_resample(self):
         # GH3609
         s = Series(range(100), index=date_range(
-            '20130101', freq='s', periods=100), dtype='float')
+            '20130101', freq='s', periods=100, name='idx'), dtype='float')
         s[10:30] = np.nan
         expected = Series([34.5, 79.5], index=[Period(
```
Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: directly create the ``PeriodIndex``

File: pandas/tseries/tests/test_timeseries.py (hunk 2 above)
Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: add the issue number (or in this case the PR number as you are fixing something unrealted to the original issue)

File: pandas/tseries/tests/test_timeseries.py [hunk 6]
```diff
@@ -2249,15 +2245,26 @@ def test_concat_datetime_datetime64_frame(self):
...
             '2013-01-01 00:00', 'T'), Period('2013-01-01 00:01', 'T')])
+        expected.index.name = 'idx'
         result = s.to_period().resample('T', kind='period').mean()
         assert_series_equal(result, expected)
         result2 = s.resample('T', kind='period').mean()
         assert_series_equal(result2, expected)
 
+    def test_empty_period_index_resample(self):
+
+        index = PeriodIndex(start='2000', periods=0, freq='D', name='idx')
```
Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: don't just compare the indexes, rather construct the expected series

File: pandas/tseries/resample.py (hunk 1 above)
Comment (MaximilianR) [inline pandas/tseries/resample.py]: Am going to punt this to https://github.com/pydata/pandas/issues/12774, and loosen the associated test in this PR

**Comment by jreback** [comment]
run ``git diff master | flake8 --diff`` to see PEP issues

**Comment by MaximilianR** [comment]
@jreback green

**Comment by MaximilianR** [comment]
Although codecov states that this line isn't tested: . Unless I'm mistaken, I added a test for this

File: pandas/tseries/tests/test_timeseries.py [hunk 7]
```diff
@@ -2249,15 +2245,27 @@ def test_concat_datetime_datetime64_frame(self):
...
         assert_series_equal(result2, expected)
 
+    def test_empty_period_index_resample(self):
+        # GH12771
+        index = PeriodIndex(start='2000', periods=0, freq='D', name='idx')
+        s = Series(index=index)
+        result = s.resample('M').sum().index
+
+        # would like to test vs the actual index, but due to GH12774,
+        # currently only testing the length
```
Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: no create the expected Series as it comes back
you can fix it later in the other issue

**Comment by jreback** [comment]
no I just fixed something on codecov

File: pandas/tseries/tests/test_timeseries.py [hunk 8]
```diff
@@ -2249,15 +2245,27 @@ def test_concat_datetime_datetime64_frame(self):
...
         assert_series_equal(result, expected)
         result2 = s.resample('T', kind='period').mean()
         assert_series_equal(result2, expected)
 
+    def test_empty_period_index_resample(self):
+        # GH12771
+        index = PeriodIndex(start='2000', periods=0, freq='D', name='idx')
+        s = Series(index=index)
+        result = s.resample('M').sum()
+        # after GH12774 is resolved, this should be a PeriodIndex
```
Comment (jreback) [inline pandas/tseries/tests/test_timeseries.py]: great. put a pointer to this PR in that issue (and maybe a checkbox), so we don't forget.

**Comment by jreback** [comment]
looks good. ping on green.

File: pandas/tseries/tests/test_timeseries.py (hunk 8 above)
Comment (MaximilianR) [inline pandas/tseries/tests/test_timeseries.py]: done

**Comment by MaximilianR** [comment]
@jreback green, apart from the codecov anomaly

**Comment by jreback** [comment]
thanks @MaximilianR 

I moved the tests to the proper location (and some other ones, not sure why there were there). ``test_resample``.

now, if in ``test_resample`` you do this (and do it with ``_ts`` as well).

```
def _simple_pts(start, end, freq='D'):
    rng = period_range(start, end, freq=freq, name='foo')
    return Series(np.random.randn(len(rng)), index=rng)
```
you will get some failures, could just be some test failures though (e.g. the expected need to change).

**Pull request closed without merging**